File: src/platforms/windows.js
const { exec } = require('child_process');
const util = require('util');

const execPromise = util.promisify(exec);

class WindowsFinder {
  async findByPort(port) {
    const { stdout } = await execPromise(`netstat -ano`);
    const processes = [];
    const lines = stdout.split('\n');
    const portStr = ':' + port;

    for (const line of lines) {
      if (!line.includes(portStr) || !line.includes('LISTENING')) continue;

      const parts = line.split(/\s+/).filter(p => p);
      if (parts.length < 5) continue;

      const pid = parseInt(parts[parts.length - 1], 10);
      if (isNaN(pid)) continue;

      const [name, user, cmd] = await this.getProcessDetails(pid);
      processes.push({
        pid,
        name: name || 'unknown',
        port,
        user: user || 'unknown',
        cmd: cmd || ''
      });
    }
    return processes;
  }

  async findByPattern(pattern) {
    const { stdout } = await execPromise(`tasklist /FO LIST /NH`);
    const processes = [];
    const re = new RegExp(pattern, 'i');
    const lines = stdout.split('\n');

    for (const line of lines) {
      const fields = line.trim().split(/\s+/);
      if (fields.length < 2) continue;

      const name = fields[0];
      if (!re.test(name)) continue;

      const pid = parseInt(fields[1], 10);
      if (isNaN(pid)) continue;

      const [, user, cmd] = await this.getProcessDetails(pid);
      processes.push({
        pid,
        name,
        user: user || 'unknown',
        cmd: cmd || ''
      });
    }
    return processes;
  }

  async listAll() {
    const { stdout } = await execPromise(`netstat -ano`);
    const processes = [];
    const lines = stdout.split('\n');
    const seen = new Set();

    for (const line of lines) {
      if (!line.includes('LISTENING')) continue;

      const parts = line.split(/\s+/).filter(p => p);
      if (parts.length < 5) continue;

      const localAddr = parts[1];
      const portMatch = localAddr.match(/:(\d+)$/);
      const port = portMatch ? parseInt(portMatch[1], 10) : 0;

      const pid = parseInt(parts[parts.length - 1], 10);
      if (isNaN(pid) || seen.has(pid)) continue;
      seen.add(pid);

      const [name, user, cmd] = await this.getProcessDetails(pid);
      processes.push({
        pid,
        name: name || 'unknown',
        port,
        user: user || 'unknown',
        cmd: cmd || ''
      });
    }
    return processes;
  }

  async getProcessDetails(pid) {
    try {
      const { stdout } = await execPromise(`wmic process where ProcessId=${pid} get Name,CommandLine /format:list`);
      let name = 'unknown';
      let cmd = '';

      const lines = stdout.split('\n');
      for (const line of lines) {
        if (line.startsWith('Name=')) {
          name = line.substring(5).trim();
        } else if (line.startsWith('CommandLine=')) {
          cmd = line.substring(12).trim();
        }
      }

      return [name, '', cmd];
    } catch {
      return ['unknown', 'unknown', ''];
    }
  }
}

class WindowsKiller {
  kill(pid, force) {
    const flag = force ? '/F' : '';
    return execPromise(`taskkill ${flag} /PID ${pid}`);
  }

  async killAll(pids, force) {
    for (const pid of pids) {
      await this.kill(pid, force);
    }
  }
}

module.exports = { WindowsFinder, WindowsKiller };

File: internal/finder/windows.go
package finder

import (
	"bufio"
	"context"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
)

type windowsFinder struct{}

func (f *windowsFinder) FindByPort(ctx context.Context, port int) ([]ProcessInfo, error) {
	// Run netstat to find processes on the port
	cmd := exec.CommandContext(ctx, "netstat", "-ano")
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	var processes []ProcessInfo
	portStr := ":" + strconv.Itoa(port)

	scanner := bufio.NewScanner(strings.NewReader(string(output)))
	for scanner.Scan() {
		line := scanner.Text()
		// Look for lines with the port and LISTENING state
		if strings.Contains(line, portStr) && strings.Contains(line, "LISTENING") {
			parts := strings.Fields(line)
			if len(parts) >= 5 {
				pid, err := strconv.Atoi(parts[len(parts)-1])
				if err != nil {
					continue
				}

				// Get process name via tasklist
				name, user, cmdLine, err := getProcessDetails(pid)
				if err != nil {
					name = "unknown"
					user = "unknown"
					cmdLine = ""
				}

				processes = append(processes, ProcessInfo{
					PID:  pid,
					Name: name,
					Port: port,
					User: user,
					CMD:  cmdLine,
				})
			}
		}
	}

	return processes, nil
}

func (f *windowsFinder) FindByPattern(ctx context.Context, pattern string) ([]ProcessInfo, error) {
	cmd := exec.CommandContext(ctx, "tasklist", "/FO", "LIST", "/NH")
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	var processes []ProcessInfo
	re := regexp.MustCompile(`(?i)` + pattern)

	scanner := bufio.NewScanner(strings.NewReader(string(output)))
	for scanner.Scan() {
		line := scanner.Text()
		fields := strings.Fields(line)
		if len(fields) >= 2 && re.MatchString(fields[0]) {
			pid, _ := strconv.Atoi(fields[1])
			_, user, cmdLine, _ := getProcessDetails(pid)
			processes = append(processes, ProcessInfo{
				PID:  pid,
				Name: fields[0],
				User: user,
				CMD:  cmdLine,
			})
		}
	}

	return processes, nil
}

func (f *windowsFinder) ListAll(ctx context.Context) ([]ProcessInfo, error) {
	cmd := exec.CommandContext(ctx, "netstat", "-ano")
	output, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	seen := make(map[int]bool)
	var processes []ProcessInfo

	scanner := bufio.NewScanner(strings.NewReader(string(output)))
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "LISTENING") {
			parts := strings.Fields(line)
			if len(parts) >= 5 {
				localAddr := parts[1]
				pid, err := strconv.Atoi(parts[len(parts)-1])
				if err != nil || seen[pid] {
					continue
				}
				seen[pid] = true

				// Extract port from local address
				port := 0
				if idx := strings.LastIndex(localAddr, ":"); idx >= 0 {
					port, _ = strconv.Atoi(localAddr[idx+1:])
				}

				name, user, cmdLine, _ := getProcessDetails(pid)
				processes = append(processes, ProcessInfo{
					PID:  pid,
					Name: name,
					Port: port,
					User: user,
					CMD:  cmdLine,
				})
			}
		}
	}

	return processes, nil
}

func getProcessDetails(pid int) (name, user, cmdLine string, err error) {
	// Get process name and command line from wmic
	cmd := exec.Command("wmic", "process", "where", "ProcessId="+strconv.Itoa(pid), "get", "Name,CommandLine", "/format:list")
	output, err := cmd.Output()
	if err != nil {
		return "unknown", "unknown", "", err
	}

	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		if strings.HasPrefix(line, "Name=") {
			name = strings.TrimPrefix(line, "Name=")
		} else if strings.HasPrefix(line, "CommandLine=") {
			cmdLine = strings.TrimPrefix(line, "CommandLine=")
		}
	}

	if name == "" {
		name = "unknown"
	}

	return name, user, cmdLine, nil
}
